### ISSUE TITLE
`<NL>` and `<C-J>`  are sometimes equal and sometimes not equal

### ISSUE THREAD

**Issue by tomtomjhj** [neovim/neovim#30887]
### Problem

`<NL>` and `<C-J>` and behaves like the same key in some cases and like different keys in the other cases.

### Steps to reproduce

* Run `nvim --clean` (tested on current HEAD and 0.10.2), and then run the following command.
  ```
  nnoremap <NL> <Cmd>echom 'NL'<CR>
  ```
  If you type `<C-J>`, `NL` is printed. I believe this is expected behavior in most terminals that don't distinguish `<C-J>` and `<NL>`.
  In this sense, `<C-J>` and `<NL>` are equal.

* `keytrans()` agrees on that.
  The following command prints `<NL> <NL>`
  ```
  echo keytrans(nvim_replace_termcodes('<C-J>', v:true, v:true, v:true)) keytrans(nvim_replace_termcodes('<NL>', v:true, v:true, v:true))
  ```

* But the problem is that you can map both `<C-J>` and `<NL>` somehow `<C-J>` is prioritized.
  Run the following commands in fresh nvim instance in that order, and type `<C-J>`.
  ```
  nnoremap <C-J> <Cmd>echom 'C-J'<CR>
  nnoremap <NL> <Cmd>echom 'NL'<CR>
  ```
  It prints `C-J`. This means that `<C-J>` and `<NL>` are actually different.
  (NOTE: In vim 9.1.771, this prints `NL`.)

Related: https://github.com/hrsh7th/nvim-cmp/pull/1935#issuecomment-2426250159

### Expected behavior

`<NL>` and `<C-J>` should be either different xor the same; not both.

### Nvim version (nvim -v)

current master and 0.10.2

### Vim (not Nvim) behaves the same?

no

### Operating system/version

ubuntu 22.04

### Terminal name/version

GNOME Terminal 3.44.0

### $TERM environment variable

tmux-256color

### Installation

manually built / appimage

**Comment by zeertzjq** [comment]
Hmm, this will require adding a new API to replace `nvim_replace_termcodes`.

**Comment by zeertzjq** [comment]
Or you can use `vim.api.nvim_eval([["\<*C-J>"]])` instead...

**Comment by justinmk** [comment]
> Or you can use `vim.api.nvim_eval([["\<*C-J>"]])` instead...

We should mention that at `:help vim.keycode` and nvim_replace_termcodes. I see that `:help expr-quote` mentions it but this needs more exposure.

**Comment by tomtomjhj** [comment]
> Surely that depends on whether modifyOtherKeys / kitty keyboard protocol is supported?

Yes. I just checked that vim distinguishes `<C-J>` and `<NL>` in kitty.

Now the real question is how and why nvim sometimes distinguishes `<C-J>` and `<NL>` inputs in gnome-terminal, while on the other hand vim doesn't distinguish them in gnome-terminal.

In any case, the original motivation for reporting this issue has been resolved with zeertzjq's suggestion.
